### PR TITLE
✨  Adding CycloneDX 1.4 and 1.5 reporter

### PR DESCRIPTION
**Pull request by marcwieserdev**
## About this PR

This PR aims to add the CycloneDX 1.4 and 1.5 format. The idea behind it is to be able to export everything in a commonly used format to later reuse the result to enrich it. (ex: re-running a vulnerability matching process on the packages to find emerging vulnerabilities).

This PR adds 2 versions of CycloneDX even though at the moment they report the exact same thing. 

In future PR, we intend to extract where packages has been found during the extraction process (file, line, column), which is an information only supported by CycloneDX 1.5.

We need a CycloneDX 1.4 format for backward compatibility reasons.

## Additional informations

You may want to review this PR commit by commit, as it should be easier to read

### CODE REVIEW

File: internal/utility/purl/composer_extractor.go [hunk 1]
```diff
@@ -0,0 +1,21 @@
+package purl
+
+import (
+	"log"
+	"strings"
+
+	"github.com/google/osv-scanner/pkg/models"
+)
+
+func ExtractPURLFromComposer(packageInfo models.PackageInfo) (namespace string, name string, ok bool) {
```
Comment (oliverchang) [inline internal/utility/purl/composer_extractor.go:10]: nit: `Extract` is a bit overloaded for us -- we already use it in the context of "extracting" dependencies from e.g. a lockfile. 

Could we just omit the `extract` bit in this directory? And further following [Go conventions to avoid repetition](https://google.github.io/styleguide/go/best-practices#avoid-repetition), could we just call these functions `FromComposer`, `FromMaven` etc? Then users will just be calling `purl.FromComposer`, `purl.FromMaven`.

File: internal/utility/purl/composer_extractor.go [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+	nameParts := strings.Split(packageInfo.Name, "/")
+	if len(nameParts) != 2 {
+		log.Printf("invalid packagist package_name=%s", packageInfo.Name)
```
Comment (oliverchang) [inline internal/utility/purl/composer_extractor.go:13]: can we return an error instead for all of these? we can replace the `ok bool` with an `err error`.

File: internal/utility/purl/golang_extractor.go [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+	"log"
+	"strings"
+
+	"github.com/google/osv-scanner/pkg/models"
+)
+
+func ExtractPURLFromGolang(packageInfo models.PackageInfo) (namespace string, name string, ok bool) {
+	nameParts := strings.Split(packageInfo.Name, "/")
+	if len(nameParts) == 0 || len(packageInfo.Name) == 0 {
+		log.Printf("invalid golang package_name=%s", packageInfo.Name)
```
Comment (oliverchang) [inline internal/utility/purl/golang_extractor.go:13]: same comment here generally: can we return errors instead of just printing and exiting?

File: internal/utility/purl/golang_extractor.go [hunk 4]
```diff
@@ -0,0 +1,25 @@
+package purl
+
+import (
+	"log"
+	"strings"
+
+	"github.com/google/osv-scanner/pkg/models"
+)
+
+func ExtractPURLFromGolang(packageInfo models.PackageInfo) (namespace string, name string, ok bool) {
```
Comment (oliverchang) [inline internal/utility/purl/golang_extractor.go:10]: nit: can we just call this `FromGo`?

File: pkg/grouper/grouper_test.go [hunk 5]
```diff
@@ -1,83 +1,82 @@
-package grouper_test
+package grouper
```
Comment (oliverchang) [inline pkg/grouper/grouper_test.go:1]: Are these changes necessary for this PR?

File: pkg/grouper/package_grouper.go [hunk 6]
```diff
@@ -0,0 +1,45 @@
+package grouper
+
+import (
+	"github.com/google/osv-scanner/internal/utility/purl"
+	"github.com/google/osv-scanner/pkg/models"
+	"slices"
+)
+
+func GroupByPURL(packageSources []models.PackageSource) map[string]models.PackageVulns {
```
Comment (oliverchang) [inline pkg/grouper/package_grouper.go:9]: Can you add some function level comments explaining what this function is doing?

File: pkg/osvscanner/vulnerability_result.go [hunk 7]
```diff
@@ -110,5 +110,6 @@ func buildVulnerabilityResults(
 		results.ExperimentalAnalysisConfig.Licenses.Allowlist = allowlist
 	}
 
+	results.ResultsByPURL = grouper.GroupByPURL(results.Results)
```
Comment (oliverchang) [inline pkg/osvscanner/vulnerability_result.go:113]: Can we avoid storing this, and just compute these on demand when this is needed?

File: pkg/grouper/package_grouper.go (hunk 6 above)
Comment (oliverchang) [reply pkg/grouper/package_grouper.go:9]: Also side comment: `grouper` is likely not the right package for this.. @another-rex WDYT?

**Review by oliverchang — COMMENTED** [review]
thank you very much for contributing this! we'd love to support this feature in OSV-Scanner. I just have some comments.

File: pkg/grouper/package_grouper.go (hunk 6 above)
Comment (marcwieserdev) [reply pkg/grouper/package_grouper.go]: Thanks for the review ! :) 
What package would you recommend ? (I thought of the `internal/utility/purl` package and then renaming the method to `Group`)

Comment (oliverchang) [reply pkg/grouper/package_grouper.go]: thanks for addressing the comments! yep `internal/utility/purl` and renaming the method to `Group` sounds great to me!

Comment (marcwieserdev) [reply pkg/grouper/package_grouper.go]: I'm on holiday right know, I will do it right when I come back (later this week)

File: go.mod [hunk 8]
```diff
@@ -26,6 +26,7 @@ require (
 	github.com/package-url/packageurl-go v0.1.3
 	github.com/pandatix/go-cvss v0.6.2
 	github.com/spdx/tools-golang v0.5.4
+	github.com/stretchr/testify v1.9.0
```
Comment (another-rex) [inline go.mod:29]: These 3 new dependencies seem to be only used for one assert, can it be done manually instead?

**Review by another-rex — COMMENTED** [review]
Thanks! I haven't had time to look into this recently, but I'll take a deeper look this week.

File: pkg/grouper/package_grouper.go (hunk 6 above)
Comment (marcwieserdev) [reply pkg/grouper/package_grouper.go]: I moved the package there, I'm resolving the conversation, feel free to open it back if anything looks off ;)

File: pkg/reporter/sbom/models.go [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+)
+
+type CycloneDXVersion int
+
+const (
+	CycloneDXVersion14 CycloneDXVersion = iota
+	CycloneDXVersion15
+)
+
+var SpecVersionToBomCreator = map[CycloneDXVersion]BomCreator{
```
Comment (another-rex) [inline pkg/reporter/sbom/models.go:16]: We will wish to add SPDX support in the future as well, can you rename this to CycloneDXBomCreator?

**Review by another-rex — COMMENTED** [review]
Some minor comments, otherwise LGTM!

**Review by another-rex — APPROVED** [review]
Thanks! Will aim to get this merged in next week.
